rbd-mirror A/A: track images in policy map

vshankar opened the pull request:
Part of tracker: http://tracker.ceph.com/issues/18786

vshankar commented:
#15691

vshankar commented:
items that need to be addressed:
* persist map changes to disk
* shuffle/unmap policy APIs

dillaman commented on src/tools/rbd_mirror/image_map/ImageSpec.h:
```diff
@@ -0,0 +1,49 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
On second glance -- is this class still used?

vshankar replied:
stale -- removed.

dillaman commented on src/tools/rbd_mirror/image_map/InstanceMapGetRequest.h:
```diff
@@ -0,0 +1,51 @@
+namespace image_map {
+
+template<typename ImageCtxT = librbd::ImageCtx>
+class InstanceMapGetRequest {
```
Nit: it's already in the "image_map" namespace, so perhaps just simplify to ```LoadRequest```?

dillaman commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,73 @@
+#include "tools/rbd_mirror/types.h"
+#include "include/rados/librados.hpp"
+
+using librados::IoCtx;
```
Never use ```using``` in a header

dillaman commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,73 @@
+  void init(Context *on_finish);
+
+  // handle additional/removal of images
+  MapSpecRef handle_images(
```
This really should be returning ```void``` since you will need an async callback to a listener.

dillaman replied:
Nit: ```handle_images``` -> ```update_images``` or ```update``` (the ```handle_``` implies it's invoked as part of a callback to me)

dillaman commented on src/tools/rbd_mirror/image_map/types.h:
```diff
@@ -0,0 +1,161 @@
+namespace mirror {
+namespace image_map {
+
+enum MapShuffleType {
```
Nit: probably can drop the ```Map``` prefix since you are in the ```image_map``` namespace

dillaman commented on src/tools/rbd_mirror/image_map/types.h:
```diff
@@ -0,0 +1,161 @@
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
+};
+
+struct MappedImage {
```
Nit: probably can drop the ```Mapped``` prefix since you are in the ```image_map``` namespace

dillaman commented on src/tools/rbd_mirror/image_map/types.h:
```diff
@@ -0,0 +1,161 @@
+
+struct MappedImage {
+  std::string id;
+  std::string global_id;
```
Nit: ```global_image_id```

dillaman commented on src/tools/rbd_mirror/image_map/types.h:
```diff
@@ -0,0 +1,161 @@
+};
+
+struct MappedImage {
+  std::string id;
```
What is this id used for? It's not obvious w/o tracing back through the code. As I mentioned in the (obsolete) ```ImageSpec``` class, I think you really need to track a set of peer UUIDs (black for local pool, populated UUID for remote peer pools). Once you drop to zero UUIDs, you know you can unmap the image from an instance.

vshankar replied:
The `id` here is not for tracking. This structure is used by `PoolReplayer::handle_update()` and `id` gets assigned to whatever we get in image_id in `PoolWatcher` (although as you mentioned that passing `id` to start image replayers is not really required as we already pass the `global_id`). This can be taken away if required.

dillaman commented on src/tools/rbd_mirror/image_map/types.h:
```diff
@@ -0,0 +1,161 @@
+  }
+};
+
+struct RemappedImage {
```
Why not just track this all within the same data structure above? It would have a state of associating, unassociating, associated, and rebalancing. Probably should also track "next instance id" for the case of rebalancing so that your policy can re-execute while a rebalance is in-progress and make decisions based on the expected state, not current state. Also should add a "last mapped" timestamp to ensure you don't ping-pong images back-and-force (i.e. a given image that is already mapped cannot be re-mapped for X minutes -- config option).

vshankar replied:
These structures are used to pass a bunch of images to `PoolReplayer::handle_update()`. Tracking next instance id (plus other things) should mostly be a part of some other "long lived" data structure such as the map itself (or a part of the map which tracks peer uuids).

Regarding the `RemappedImage` structure, yeh, we have just stick to having just one structure for representing mapped, removed and remapped images.

dillaman commented on src/tools/rbd_mirror/image_map/types.h:
```diff
@@ -0,0 +1,161 @@
+  std::ostream &, const MapSpec &mapspec);
+
+// in-memory map of tracked images
+struct InImage {
```
I think you would only need this data structure on a transient basis when loading the map from disk and then the policy's in-memory map would be initialized from it. Of course, the global image id -> instance id map is so simple it could just be represented by a std::map

dillaman commented on src/cls/rbd/cls_rbd_types.h:
```diff
@@ -364,6 +364,34 @@ struct TrashImageSpec {
 };
 WRITE_CLASS_ENCODER(TrashImageSpec);
 
+enum MirrorImageMapState {
```
Let's drop this for now -- right now, the existence of the struct indicates the image is associated (or dissociating or associating).

vshankar replied:
Done. I've taken out the state tracking part and just maintain `instance_id` on disk.

vshankar commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -3097,6 +3097,7 @@ static const std::string IMAGE_KEY_PREFIX("image_");
 static const std::string GLOBAL_KEY_PREFIX("global_");
 static const std::string STATUS_GLOBAL_KEY_PREFIX("status_global_");
 static const std::string INSTANCE_KEY_PREFIX("instance_");
+static const std::string MIRROR_IMAGE_MAP_KEY_PREFIX("map_image_");
```
Wouldn't that result in these entries being picked by when `IMAGE_KEY_PREFIX` is used as the key? I had faced this earlier...

dillaman replied:
Are they used in the same object? I thought one was in rbd_mirroring and the other was in rbd_leader? Could also name it "instance_image_"

vshankar replied:
In the sense, I used `instance_map` earlier and that got these keys to be listed out where `INSTANCE_KEY_PREFIX` was being used (leader object). So I stuck to keeping it distinct. `image_map` would be safe to use here.

vshankar commented:
updated pr with a minor chage in tracking async operations in policy.

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -3684,6 +3689,84 @@ int instances_remove(cls_method_context_t hctx, const string &instance_id) {
+
+    max_read = MIN(RBD_MAX_KEYS_READ, max_return - image_map->size());
+    if (!vals.empty()) {
+      last_read = mirror_image_map_key(image_map->rbegin()->first);
```
The key should already be good-to-go -- no need to prefix an additional prefix onto it.

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -3684,6 +3689,84 @@ int instances_remove(cls_method_context_t hctx, const string &instance_id) {
+      const std::string& global_image_id =
+        it->first.substr(MIRROR_IMAGE_MAP_KEY_PREFIX.size());
+
+      cls::rbd::MirrorImageMap imagemap;
```
Nit: ```image_map```

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -3684,6 +3689,84 @@ int instances_remove(cls_method_context_t hctx, const string &instance_id) {
+  int max_read = MIN(RBD_MAX_KEYS_READ, max_return);
+  std::string last_read = mirror_image_map_key(start_after);
+
+  while (max_read) {
```
Nit: ```max_read > 0```

dillaman commented on src/cls/rbd/cls_rbd.cc:
```diff
@@ -3684,6 +3689,84 @@ int instances_remove(cls_method_context_t hctx, const string &instance_id) {
+int mirror_image_map_list(cls_method_context_t hctx,
+                          const std::string &start_after,
+                          uint64_t max_return,
+                          std::map<std::string, cls::rbd::MirrorImageMap> *image_map) {
```
Nit: ```image_maps``` or ```image_mapping```

dillaman commented:
@vshankar ignore -- apparently I was looking at the wrong PR

dillaman commented on src/tools/rbd_mirror/image_map/Types.h:
```diff
@@ -0,0 +1,97 @@
+enum ShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
```
Nit: case doesn't appear to be used (of course, not sure the ```ShuffleType``` would need to live either)

dillaman commented on src/tools/rbd_mirror/image_map/SimplePolicy.cc:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  uint64_t images_per_instance = nr_images / nr_instances;
+  if (nr_images % nr_instances) {
```
Nit: not a boolean result so append w/ ``` != 0```

dillaman commented on src/tools/rbd_mirror/image_map/SimplePolicy.cc:
```diff
@@ -0,0 +1,186 @@
+
+bool instance_present(const std::vector<std::string> &instance_ids,
+                      const std::string &instance) {
+  return !(std::find(instance_ids.begin(),
```
Nit: swap ```==``` for ```!=``` and drop the negation

vshankar commented on src/tools/rbd_mirror/image_map/ImageAction.h:
```diff
@@ -0,0 +1,138 @@
+public:
+  ImageAction();
+
+  enum MirrorPeerType {
```
I kind of found it easier when local/remote images maintain (an follow) their respective state transitions -- although for the general add/remove case its pretty much straight forward (i.e., no differentiation), I found it slightly convenient for the shuffle case.

vshankar replied:
Or are you suggesting that there's no point in maintaining current state for images?

dillaman replied:
You only need to track actions at the global image id level -- you shouldn't care about state transitions for individual mirror peer images. That is why I have a comments below that you should track images using a "global image id -> set<mirror uuid>" mapping, that you should map an image once it is first added, and you should unmap and image once all mirror peers (remote or local) have been removed from said mapping.

vshankar commented on src/tools/rbd_mirror/image_map/ImageAction.h:
```diff
@@ -0,0 +1,138 @@
+  Peer &peer_type(ActionState &action_state, MirrorPeerType peer_type);
+
+  ImageAction::State current_state(const Peer &peer);
+  ImageAction::State reach_state(Peer &peer);
```
yeh -- done.

vshankar commented on src/tools/rbd_mirror/image_map/ImageAction.cc:
```diff
@@ -0,0 +1,362 @@
+  // ACTION_ADD
+  {
+    {SUB_ACTION_MAP,      STATE_ASSOCIATED},    // STATE_UNASSIGNED
+    {SUB_ACTION_COMPLETE, STATE_ASSOCIATED},    // STATE_ASSOCAITED
```
oops, thanks!

vshankar commented on src/tools/rbd_mirror/image_map/RemoveRequest.h:
```diff
@@ -0,0 +1,62 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
It kind of felt a bit odd to accept a map of `images -> data` to be updated and a set of image ids to remove, so I made them into separate state machines (earlier I did have them clubbed up as they were accepting a map of `image -> instance ids` with empty instance_id hinting removal).

vshankar commented on src/tools/rbd_mirror/image_map/ImageAction.cc:
```diff
@@ -0,0 +1,362 @@
+  },
+  // ACTION_REMOVE
+  {
+    {SUB_ACTION_UNMAP,    STATE_DISASSOCIATED}, // STATE_UNASSIGNED
```
The map could still be present on-disk. We should jump to `PURGING` state in that case. I do not clearly recall why I followed the normal removal state transition when the state is `UNASSIGNED`.

vshankar commented:
minor update: added encoder test for `rbd::mirror::image_map::MirrorPolicyData`

dillaman commented on src/tools/rbd_mirror/image_map/ImageAction.h:
```diff
@@ -0,0 +1,122 @@
+
+    SubActionType sub_action_type; // currently executing sub action
+
+    Action(ActionType action_type)
```
Nit: recommend moving this to private and creating three static factory methods for each action (i.e. ```create_remove```). Each factory method would take the params for its specific callbacks to be more explicit about what is needed.

vshankar replied:
Make sense. In that case, do you have a preference if the returned object needs to be dynamically allocated to avoid unnecessary copying (one during the return, which mostly should be optimized by the compiler and another when the action is queued)?

`struct Action` is not huge for the copy to be expensive, so I'm guessing its ok.

dillaman replied:
No need for allocating on the heap -- this is a tiny object. You can always use move semantics to transfer ownership. Of course, if the callbacks are going to be consolidated down to a fixed number for all actions due to other changes, perhaps this change isn't necessary.

dillaman commented on src/tools/rbd_mirror/image_map/ImageAction.h:
```diff
@@ -0,0 +1,122 @@
+  struct Action {
+    ActionType action_type;
+
+    Context *on_map;
```
Nit: map of ```SubActionType``` to ```Context *``` for simplicity.

dillaman commented on src/tools/rbd_mirror/image_map/ImageAction.h:
```diff
@@ -0,0 +1,122 @@
+  };
+
+  enum SubActionType {
+    SUB_ACTION_TYPE_INVALID = 0,
```
Nit: can we drop this and just use ```COMPLETE```?

vshankar replied:
yeh ok -- this will go away with the `assert` during invalid transition.

dillaman commented on src/tools/rbd_mirror/image_map/ImageAction.cc:
```diff
@@ -0,0 +1,300 @@
+
+  derr << ": invalid transition (action_type=" << action_type << ", state="
+       << state << ")" << dendl;
+  return {SUB_ACTION_TYPE_INVALID, STATE_UNASSIGNED};
```
Nit: assert?

dillaman commented on src/tools/rbd_mirror/image_map/ImageAction.h:
```diff
@@ -0,0 +1,122 @@
+    }
+  };
+
+  Mutex m_lock;
```
Is the lock required? Looks like the ```Policy::m_map_lock``` is always locked when using the ```ImageAction``` instance

vshankar replied:
Technically, its not required. But I'm thinking if at all locking in `ImageAction` is required in future. Also, this lock is never contented as of now (due to all access is under `Policy::m_map_lock`), so its pretty much lightweight.

dillaman replied:
Personally -- I'd yank it since ```ImageAction``` should always be used under the image map's lock

dillaman commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,684 @@
+    RWLock::WLocker map_lock(m_map_lock);
+
+    for (auto const &global_image_id : global_image_ids) {
+      ImageAction::SubActionType sub_action_type = m_image_action.sub_action_start(global_image_id);
```
Invoking this method will invoke one of the subaction callbacks -- might be nice if the switch below could be eliminated and instead have those callbacks queue themselves in the set/remove update lists.

dillaman commented on src/common/options.cc:
```diff
@@ -5276,6 +5276,14 @@ static std::vector<Option> get_rbd_mirror_options() {
     .set_default(3)
     .set_description("number of failed attempts to acquire lock after missing heartbeats before breaking lock"),
+
+    Option("rbd_mirror_image_policy_update_check_interval", Option::TYPE_INT, Option::LEVEL_ADVANCED)
```
Nit: maybe a double instead of int so that you can advance the state machine faster?

dillaman commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,251 @@
+      if (r == 0) {
+        policy->notify_listener(to_map, to_unmap, to_purge);
+      } else {
+        // on failure, resume pending actions (if any) for this
```
On failure I don't think it would be safe to proceed to the action. If you failed to notify an instance, try again until that instance is removed (marked dead) so that you know it's safe to proceed.

dillaman commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,684 @@
+  // on an ack, finish the current sub-action and schedule the
+  // next if available or start a next action if finished.
+  // on a nack: directly finish the current action (not that
+  // this will invoke the context callback if any with a -ve
```
Nit: ```-ve``` -> ```negative```

vshankar commented on src/common/options.cc:
```diff
@@ -5276,6 +5276,14 @@ static std::vector<Option> get_rbd_mirror_options() {
+    .set_default(1)
+    .set_description("interval (in seconds) to check images for mirror daemon peer updations"),
+
+    Option("rbd_mirror_image_policy_migration_threshold", Option::TYPE_INT, Option::LEVEL_ADVANCED)
```
`throttle` doesn't look accurate to me -- how about `rbd_mirror_image_policy_migration_interval`

dillaman replied:
I only suggested throttle since it's what prevents the image from migrating more than X times per Y. To me, an interval implies that is will migrate every X seconds.

vshankar commented on src/tools/rbd_mirror/CMakeLists.txt:
```diff
@@ -21,6 +22,12 @@ set(rbd_mirror_internal
   image_map/LoadRequest.cc
   image_map/UpdateRequest.cc
+  image_map/ImageAction.cc
+  image_map/LoadRequest.cc
```
shoot -- that was caused by a rebase after the cls/rbd methods got merged. Fixed (had duplicate `LoadRequest` and `UpdateRequest` entries too).

vshankar commented on src/tools/rbd_mirror/image_map/Types.h:
```diff
@@ -0,0 +1,91 @@
+
+struct MirrorPolicyData {
+  MirrorPolicyData()
+    : policy_meta(MirrorPolicyMetaTypeNone()) {
```
why `Unknown`? wouldn't we want to have policy data empty now and later introduce tracking variables to decide what and where to migrate? Or do you mean to get rid to `None` policy meta type and just have `Unknown` for now (which as with this version of the pr has an `assert` placed deliberately in its `encode`)

dillaman replied:
I would like to force the internal code users to explicitly pick what they are choosing to store instead of allowing it to default to a valid meta type. If the simply policy doesn't override it w/ the none meta, that should be considered a bug.

vshankar commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,684 @@
+        ctx->to_map.emplace_back(global_image_id, instance_id);
+        break;
+      case ImageAction::SUB_ACTION_TYPE_UNMAP:
+        ctx->to_unmap.emplace_back(global_image_id, instance_id);
```
good point about skipping the on-disk update for unmap. will fix this accordingly.

vshankar commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,251 @@
+
+    Updates to_map;
+    Updates to_unmap;
+    Updates to_purge;
```
yeh, with the remove image/unmap refactor, this would go away.

vshankar commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,684 @@
+
+    if (m_image_info[global_image_id].registered_peers.empty()) {
+      unmap(global_image_id);
+      schedule_remove_action(global_image_id, instance_id,
```
Just checking back on this -- we do not need to wait for (all) removal notification acks from the mirror daemon peer before scheduling an `unmap` action, right? We'll be firing off the removal notifications, and on the last remove (empty peers) schedule a remove action (`unmap + purge`). This would maintain the order or `remove` and `release`.

dillaman replied:
Assuming ```registered_peers``` contains both the remote peers and the local "peer", there shouldn't be a need to wait. If the last peer is removed it implies that the local image has successfully been deleted.

vshankar replied:
yup, thats what it is.

vshankar commented:
@dillaman couldn't write up a list of changes in this current version (from the last version) of the PR.

vshankar commented on src/tools/rbd_mirror/ImageMap.h:
```diff
@@ -0,0 +1,277 @@
+  // retry friendly callback contexts for sub actions. note that
+  // callback context when an action is completed need not be
+  // of this type as its just invoked once, upon completion.
+  struct SubActionContext : Context {
```
I feel there could be a better way of implementing this abstract class and how its used by the derived classes below (`SubActionMapContext` etc..). Thoughts?

vshankar commented:
changes in the latest update (w/ comments incorporated)

* split the original `image_map::Policy` class into `ImageMap` and `image_map::Policy`. `ImageMap` schedules image updates, persists on-disk image map, notifies listener and makes use of `image_map::Policy`.
* `image_map::Policy` is an abstract class which maintains the master instance map (`instance_id` -> `images`) and provides public APIs to `lookup`, `map`, `unmap` and `shuffle` (to move images mapped to an instance to another instance). Implementations of `image_map::Policy` need to provide custom implementations of a handful of pure virtual functions. These are invoke by `image_map::Policy` as per the need. `image_map::SimplePolicy` is such an implementation of this abstract class which does a simple mapping of M images to N instances.

dillaman commented on src/tools/rbd_mirror/image_map/SimplePolicy.cc:
```diff
@@ -0,0 +1,171 @@
+
+template <typename I>
+void SimplePolicy<I>::do_shuffle_remove_instances(const std::vector<std::string> &instance_ids,
+                                                  std::map<std::string, std::pair<std::string, std::string> > *remap) {
```
Nit: line length

dillaman commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,96 @@
+  // policy needs to implement these..
+
+  // map an image (global image id) to an instance
+  virtual std::string do_map(const std::string &global_image_id) = 0;
```
Nit: virtual methods should be ```protected``` at a minimum

vshankar replied:
I generally follow Herb Sutter's recommendation for keeping virtual functions private (unless those need to be invoked from derived class itself): http://www.gotw.ca/publications/mill18.htm

Thoughts?

dillaman replied:
I've always viewed it as "private" methods imply that only this particular class can invoke them. When you have "private" virtual methods, you are breaking that assumption since they could be invoked from a derived or base class. That is why I've always personally found it odd to ever mark virtual functions as private (especially since there is "protected" to delineate methods that are not part of the public interface but are used for functionality between the inheritance hierarchy.

dillaman commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -0,0 +1,663 @@
+    // track dead instances
+    for (auto const &instance : instances) {
+      if (!m_policy->instance_present(instance)) {
+        m_dead_instances.insert(instance);
```
I am unclear about the purpose of ```m_dead_instances```  --- it seems that it's only tracking instances that don't have images mapped to them?

dillaman commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,239 @@
+                        std::map<std::string, std::pair<std::string, std::string> > *remap_ids) {
+  dout(20) << dendl;
+
+  // for now this doesn't support updating instances with mix of
```
I'd say this should either be broken into ```add_instance``` and ```remove_instance``` methods or it should handle mixed add/remove cases. The former would probably be easier (and cleaner). This would impact ```ImageMap```'s interface as well.

dillaman commented on src/tools/rbd_mirror/ImageAction.h:
```diff
@@ -0,0 +1,124 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
I almost wonder if ```ImageAction``` should be integrated with / semi-merged into ```Policy``` since right now the in-memory policy map might not actually reflect the reality of where the current state mapping is actually applied. I think that would help eliminate some of the potential race conditions / edge conditions between these two separate image state tracking mechanisms.

vshankar replied:
`ImageAction` is merely just a state machine invoking appropriate context callbacks provided by its consumer. There's no tracking at all in terms of which instance id to is being notified for a given image and the current sub-action being executed. Did you mean parts of `ImageMap` (especially the retry logic in there) to be better integrated with `Policy`?

dillaman replied:
The policy and the state machine should be integrated together so they cannot get out-of-sync. If the policy wants to move an image to a new instance, the policy should be internally tracking the state changes of the image. Right now the in-memory policy can think and image is on instance XYZ while it actually might be still on ABC due to the state machine transition.

vshankar replied:
thanks! I did most of the changes integrating `ImageAction` and `Policy`, but then thought if you meant (by mistake) to move pieces from `ImageMap` to `Policy`. Once we start tracking state changes in `Policy` the retry logic can go away from `ImageMap`.

dillaman commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,239 @@
+namespace {
+
+bool can_shuffle_image(utime_t now, utime_t last_shuffled_time, int migration_throttle) {
+  return !(migration_throttle > 0 && (now - last_shuffled_time) < migration_throttle);
```
Related to the comment about integrating ```ImageAction``` and ```Policy``` -- we shouldn't be migrating the image if that image already has an action in-progress. The only thing that should override an in-progress action is when the peer for the in-progress destination action is removed (i.e. it's safe to remap).

vshankar replied:
checking back on this one -- for the instance removal case, we would want to restart the (pending) actions after the images (currently mapped to the removed instance) have been migrated to another mirror peer, right?

dillaman replied:
When the policy is informed that a peer has been removed, you don't need to tell that peer that it should remove an image (since the peer is dead / blacklisted). You just need to map it somewhere else. Similarly, if you have an image that is in-progress being mapped to a removed peer, you would need to pick a new peer to map it to.

dillaman commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,239 @@
+  for (auto it = (*remap_ids).begin(); it != (*remap_ids).end();) {
+    const std::string &global_image_id = (*it).first;
+
+    if (!can_shuffle_image(now, m_shuffled_timestamp[global_image_id], migration_throttle)) {
```
If the image cannot be shuffled -- the ```SimplyPolicy``` should have excluded it from its decision making

vshankar replied:
Shouldn't this override (based to last shuffled time) be decided by the policy class rather than its implementers? Here, `SimplePolicy` just returns a set of image ids based on its balancing algorithm.

dillaman replied:
Yeah -- but if this class is going to override the decision, then the policy will be sub-optimal since it might be picking an image that cannot move to move when instead it could pick an image that can move instead.

dillaman commented on src/tools/rbd_mirror/image_map/Policy.cc:
```diff
@@ -0,0 +1,239 @@
+  for (auto const &instance : instances) {
+    dout(10) << ": removing instance_id=" << instance << dendl;
+    if (!m_map[instance].empty()) {
+      dout(5) << ": instance_id=" << instance << " still has " << m_map[instance].size()
```
Would this be a bug?

vshankar replied:
It would be if the policy implementation does not (for some reason) remap all images from the instances that are being removed. I did not treat it as a hard error here. With the changes in `update_images_removed()` (ignoring if an image is not found in the map), we should probably `assert()` here.

dillaman replied:
Sounds good to me

dillaman commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,96 @@
+  librados::IoCtx &m_ioctx;
+  std::map<std::string, utime_t> m_shuffled_timestamp;
+
+  std::string _lookup(const std::string &global_image_id);
```
Nit: prefer just passing the lock as an argument to indicate that the lock is already held vice prefixing w/ ```_```

dillaman commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  // build an initial instance map
+  void build_map(const std::map<std::string, cls::rbd::MirrorImageMap> &image_mapping);
```
Nit: perhaps should be named ```init``` since it's not computing anything

dillaman commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -0,0 +1,663 @@
+
+  for (auto const &global_image_id : global_image_ids) {
+    std::string instance_id = m_policy->lookup(global_image_id);
+    assert(instance_id != Policy<I>::UNMAPPED_INSTANCE_ID);
```
Nit: probably should just continue if the instance is already unmapped. It might be possible for a duplicate remove event to fire.

dillaman commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -0,0 +1,663 @@
+
+  // removal notification will be notified instantly. this is safe
+  // even after scheduling action for images as we still hold m_lock
+  notify_listener_remove_images(mirror_uuid, to_remove);
```
Nit: wrap w/ ```if (!mirror_uuid.empty())```

dillaman commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -0,0 +1,663 @@
+           << peer_uuid << dendl;
+
+  auto rm = m_image_info[global_image_id].registered_peers.erase(peer_uuid);
+  assert(rm);
```
Nit: might be possible that peer sends duplicate removal event, so just ignore.

dillaman commented on src/tools/rbd_mirror/ImageMap.h:
```diff
@@ -0,0 +1,278 @@
+
+    void complete(int r) override {
+      finish(r);
+      image_map->set_retry_instance_id(global_image_id, instance_id);
```
I think if ```ImageAction``` and ```Policy``` are better integrated, the policy would already know where the image is currently mapped and where the desired next mapping is located. Therefore, you wouldn't need to separately track where to re-send notifications.

dillaman commented on src/tools/rbd_mirror/ImageMap.h:
```diff
@@ -0,0 +1,278 @@
+        instance_id(instance_id) {
+    }
+
+    void update_map_sub_action() {
```
Nit: why have these "helper" methods instead of just directly invoking the methods from the appropriate ```finish``` override?

vshankar replied:
No strong reason to have it this way... `SubActionContext` invokes `set_retry_instance_id()` which requires it to record `global_image_id` and `instance_id`. Alternatively, each derived class could have accessed the base class member variable (via `SubActionContext::`) and invoked the appropriate methods in `finish()`. Just found this a bit cleaner.

I'll relook at this w.r.t. your comment on integrating `Policy` and `ImageActions`.

vshankar commented on src/tools/rbd_mirror/ImageAction.h:
```diff
@@ -0,0 +1,124 @@
+
+// NOTE: this class assumes callers take care of serialization when
+// invoking class APIs
+class ImageAction {
```
done -- this and moved to `image_map`.

vshankar commented:
@dillaman PR updated with changes.

vshankar commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,185 @@
+  std::map<std::string, ActionState> m_actions;
+  std::map<std::string, utime_t> m_shuffled_timestamp;
+
+  std::set<std::string> m_dead_instances;
```
I had to retain this as policy keeps the instance around till all images mapping to an instance are unmapped (migrated) when an instance is removed.

vshankar commented on src/tools/rbd_mirror/image_map/StateTransition.h:
```diff
@@ -0,0 +1,59 @@
+  enum SubActionType {
+    SUB_ACTION_TYPE_MAP = 0,
+    SUB_ACTION_TYPE_UNMAP,
+    SUB_ACTION_TYPE_PURGE,
```
not sure I understood this fully -- would this mean having on-disk map updates and acquiring/removing images as two separate passes (two state-machine transitions)?

dillaman replied:
Yeah -- personally I think it would be clearer in that it shows the true state of the system (waiting on an on-disk update vs waiting for a instance to ACK the acquire/release). Perhaps it make sense to keep one as UPDATING_MAPPING and another as REMOVING_MAPPING if that helps get objects in the correct collection.

vshankar commented on src/tools/rbd_mirror/image_map/StateTransition.h:
```diff
@@ -0,0 +1,59 @@
+    ActionType action_type;
+    State current_state;
+    Transition transition;
+  };
```
Just for clarity -- retries with peer would be handled by the listener (of `ImageMap` class)?

dillaman replied:
Yes -- ```InstanceWatcher``` will retry the peer until successful or until the peer has been blacklisted (this last part needs to be handled).

vshankar commented on src/tools/rbd_mirror/ImageMap.cc:
```diff
@@ -0,0 +1,613 @@
+    on_map, on_unmap, on_finish);
+
+  if (is_remove) {
+    prepend_and_schedule_action(global_image_id, action);
```
The purpose of the prepend logic is to queue a shuffle action (due to instance removal) in the front the action queue. I think this is still required, right?

dillaman replied:
If you had an action in-progress, you can't really interrupt it mid-state machine for a new action. If you are trying to send an acquire or release notification to an instance that was removed, the in-flight action can be aborted via the error path handling for the current state and it can move on to the next action. The policy shouldn't really allow the actions to get more than 2 deep at maximum since the last action can always be replaced by a new action if it hasn't started yet.

vshankar commented on src/tools/rbd_mirror/image_map/Policy.h:
```diff
@@ -0,0 +1,185 @@
+
+  // shuffle images when instances are added/removed
+  void add_instances(const std::vector<std::string> &instance_ids,
+                     std::set<std::string> *remap_global_image_ids, bool shuffle);
```
actually this can be removed, but not due to the reason of it always being `false` but because the map is empty when `Policy` is instantiated -- so `do_shuffle_add_instances()` would not do anything.

vshankar commented:
changes in this update:

* addressed comments
* context callback for `on-disk map update, acquire and release` are `StateTransition::State` based
(removed `subactions` from everywhere)
* operation retry or abort is now handled by `Policy` (was handled by `ImageMap` earlier). also, only on-disk map updates are retried on failure.
* `Policy::remove_instances()` is not a virtual function anymore (and not policy implementation dependent) -- it should always remap images to other (live) instances.

vshankar commented on src/tools/rbd_mirror/ImageMap.h:
```diff
@@ -0,0 +1,260 @@
+      image_map->queue_map_sub_action_update(global_image_id);
+    }
+
+    void complete(int r) override {
```
I think this needs to be kept as it is as `r` is always 0 when this context is invoked via the state machine and on-disk map updates may be retried upon failure.

tchaikov commented:
this change breaks the "make check" on aarch64 consistently. see https://jenkins.ceph.com/job/ceph-pull-requests-arm64/9051/consoleFull#-2069175742d63714d2-c8d8-41fc-a9d4-8dee30be4c32

```
[ RUN      ] TestMockImageMap.SetLocalImages
seed 46228
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/tools/rbd_mirror/image_map/StateTransition.cc: In function 'static const rbd::mirror::image_map::StateTransition::Transition& rbd::mirror::image_map::StateTransition::transit(rbd::mirror::image_map::StateTransition::ActionType, rbd::mirror::image_map::StateTransition::State)' thread ffff666d4f40 time 2017-11-16 02:04:30.983796
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/tools/rbd_mirror/image_map/StateTransition.cc: 78: FAILED assert(false)
 ceph version 13.0.0-3210-g503ce71 (503ce7158d23a8f2da60e79690b5780bad56a9b7) mimic (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0xfc) [0xffff9915364c]
 2: (()+0x7cede8) [0xaaaad9fe6de8]
 3: (rbd::mirror::image_map::Policy::start_next_action(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xb4) [0xaaaad9fe185c]
 4: (()+0x4b9848) [0xaaaad9cd1848]
 5: (FunctionContext::finish(int)+0x3c) [0xaaaad9ce99e4]
 6: (Context::complete(int)+0x1c) [0xaaaad9ce131c]
 7: (()+0x4b4ec8) [0xaaaad9cccec8]
 8: (FunctionContext::finish(int)+0x3c) [0xaaaad9ce99e4]
 9: (Context::complete(int)+0x1c) [0xaaaad9ce131c]
 10: (ThreadPool::worker(ThreadPool::WorkThread*)+0xa44) [0xffff9915ac04]
 11: (ThreadPool::WorkThread::entry()+0x14) [0xffff9915bb74]
 12: (()+0x6fc4) [0xffffa1a6afc4]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

```

@vshankar @dillaman

vshankar commented:
I'll look into it.

tchaikov commented:
@vshankar my wild guess is the typecast in `EncodeVisitor`. i am pulling together a quick fix.

tchaikov commented on src/tools/rbd_mirror/image_map/Types.h:
```diff
@@ -0,0 +1,101 @@
+
+struct PolicyMetaUnknown {
+  static const PolicyMetaType TYPE = static_cast<PolicyMetaType>(-1);
+
```
i think this is suspicious. as the underlying type of enum is `unsigned` by default. and casting it to and from `uint32_t` does not always work.

dillaman replied:
The point here to to assign it to an unused ID (i.e. 0xFFFFFFFF)

tchaikov replied:
my concern was that the MSBs of the number could be removed when being casted. but that's not the case.
